mirrord-layer/src/pod_api.rs

```diff
@@ -412,8 +381,252 @@ async fn create_job_pod_agent(
         .items
         .first()
         .and_then(|pod| pod.metadata.name.clone())
-        .ok_or(LayerError::PodNotFound(mirrord_agent_job_name))?;
+        .ok_or(LayerError::JobPodNotFound(mirrord_agent_job_name))?;
 
     wait_for_agent_startup(pods_api, &pod_name, "mirrord-agent".to_string()).await?;
     Ok(pod_name)
 }
+
+#[derive(Debug, Clone)]
+pub(crate) struct Response {
+    container_info: Option<String>,
+    node_name: Option<String>,
+    pod_name: Option<String>,
+    container_name: Option<String>,
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub(crate) struct DeploymentData {
+    pub deployment: String,
+}
+
+impl DeploymentData {
+    pub async fn container_data(
+        &self,
+        deployment_api: &Api<Deployment>,
+        pod_api: &Api<Pod>,
+    ) -> Option<Response> {
+        let deployment = deployment_api.get(&self.deployment).await.ok()?;
+        let pod_label = deployment
+            .spec
+            .and_then(|spec| spec.template.metadata?.labels)
+            .and_then(|pod_name| pod_name.get("app").cloned())?;
+        let pod = pod_api
+            .list(&ListParams::default().labels(&format!("app={}", pod_label)))
+            .await
+            .ok()?
+            .items
+            .first()?
+            .clone();
+        let pod_name = pod.clone().metadata.name;
+        let container_info = pod
+            .clone()
+            .status?
+            .container_statuses?
+            .first()?
+            .container_id
+            .clone();
+        let container_name = pod
+            .clone()
+            .spec?
+            .containers
+            .first()
+            .map(|container| container.name.clone());
+        let node_name = pod.spec?.node_name;
+        Some(Response {
+            container_info,
+            node_name,
+            pod_name,
+            container_name,
+        })
+    }
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub(crate) enum Target {
+    Deployment(DeploymentData),
+    Pod(PodData),
+}
+
+impl Target {
+    pub async fn container_info(&self, client: &Client, namespace: &str) -> Result<Response> {
+        let (pod_api, deployment_api) = self.target_apis(client, namespace)?;
+        match self {
+            Target::Pod(pod) => pod
+                .container_data(&pod_api)
+                .await
+                .ok_or_else(|| LayerError::PodNotFound(self.clone())),
+            Target::Deployment(deployment) => deployment
+                .container_data(&deployment_api, &pod_api)
+                .await
+                .ok_or_else(|| LayerError::DeploymentNotFound(deployment.deployment.clone())),
+        }
+    }
+
+    fn target_apis(&self, client: &Client, namespace: &str) -> Result<(Api<Pod>, Api<Deployment>)> {
```

Comment:
Seems unnecessary to split it into a separate method since it's only used in one place inside the struct itself.